File: app-mobile/src/api/events.ts
/**
 * Events API (DESIGN §8.7 - "the Partiful corner"): org events, invites and RSVPs.
 * Mirrors backend routers/events.py (c33, c198).
 */

import { request } from "./client";

export type RsvpStatus = "going" | "maybe" | "cant";

/**
 * How far an event travels, narrowest first. Mirrors the events.visibility CHECK.
 *
 *   chapter  - active members of the hosting chapter only
 *   campus   - .edu-verified students of this chapter's campus
 *   verified - any .edu-verified user (sister chapter, another school)
 *   public   - anyone at all, no account needed
 *
 * An INVITE is a separate axis and admits one named person whatever this says.
 */
export type EventVisibility = "chapter" | "campus" | "verified" | "public";

export interface EventCreate {
  title: string;
  /** ISO-8601 instant. Replaced the old free-text date_label in c198. */
  starts_at: string;
  location: string;
  cover_url: string;
  description?: string | null;
  ends_at?: string | null;
  /** Omitted means 'chapter' - the server defaults to the narrowest tier, never wider. */
  visibility?: EventVisibility;
}

/**
 * A host's edit. Omitted fields are left alone (PATCH semantics); an explicit `null`
 * on ends_at or description CLEARS it (c202) - the backend's exclude_unset check
 * (routers/events.py update_event) distinguishes "key not sent" from "key sent as
 * null", so those two fields carry `| null` here where the rest do not.
 */
export interface EventUpdate {
  title?: string;
  starts_at?: string;
  ends_at?: string | null;
  location?: string;
  cover_url?: string;
  description?: string | null;
  visibility?: EventVisibility;
}

export interface EventOut {
  id: string;
  chapter_id: string;
  title: string;
  /** Picsum-seeded cover photo (§10.2 - real imagery over placeholders). */
  cover_url: string;
  description: string | null;
  starts_at: string;
  ends_at: string | null;
  location: string;
  visibility: EventVisibility;
  /** Non-null means the party is off. The row survives so guests can be told. */
  canceled_at: string | null;
  host_id: string;
  created_at: string;
}

export interface EventInviteOut {
  event_id: string;
  invited_user_id: string;
  invited_by: string;
  created_at: string;
}

/** Response of POST /events/{id}/invites (c359) - `created` is exactly the rows
 * THIS call inserted, not the event's whole accumulated invite history. Empty on
 * an all-duplicate/all-blocked/no-op batch. */
export interface EventInviteBatchOut {
  created: EventInviteOut[];
  created_count: number;
}

export interface EventRsvpOut {
  event_id: string;
  user_id: string;
  status: RsvpStatus;
  created_at: string;
}

/**
 * Cursor options for the split guest-list routes (c275). The tie-break key is a
 * USER id, not a row id - rsvp/invite rows have composite primary keys - so the
 * second cursor param carries the last row's user_id / invited_user_id.
 */
export interface GuestListPage {
  after?: string;
  afterUserId?: string;
  limit?: number;
}

export interface EventInvitePage extends GuestListPage {
  unansweredOnly?: boolean;
}

export interface EventPageResult<T, Cursor> {
  items: T[];
  /** A full page may require one final empty request to establish exhaustion. */
  next: Cursor | null;
}

export const EVENT_GUEST_PAGE_SIZE = 200;
export const HOME_INVITE_PAGE_SIZE = 50;

/** All events for a chapter (Events segment, §8.7), soonest-first by start time. */
export async function listEvents(chapterId: string): Promise<EventOut[]> {
  return request<EventOut[]>(`/chapters/${chapterId}/events`);
}

export async function createEvent(chapterId: string, body: EventCreate): Promise<EventOut> {
  return request<EventOut>(`/chapters/${chapterId}/events`, { method: "POST", body });
}

export async function getEvent(eventId: string): Promise<EventOut> {
  return request<EventOut>(`/events/${eventId}`);
}

/** This caller's answer, even when their row is beyond the first guest page. */
export async function getMyRsvp(eventId: string): Promise<{ status: RsvpStatus | null }> {
  return request<{ status: RsvpStatus | null }>(`/events/${eventId}/rsvps/mine`);
}

/** Edit an event. Host or e-board only; a cancelled event is not editable. */
export async function updateEvent(eventId: string, body: EventUpdate): Promise<EventOut> {
  return request<EventOut>(`/events/${eventId}`, { method: "PATCH", body });
}

/** Call the party off. Idempotent - the first cancellation moment is the one kept. */
export async function cancelEvent(eventId: string): Promise<EventOut> {
  return request<EventOut>(`/events/${eventId}/cancel`, { method: "POST" });
}

/**
 * Invite people. Host or e-board only, because an invite GRANTS READ ACCESS.
 *
 * Returns only the rows THIS call inserted (c359), not the event's whole
 * accumulated invite history - `created` is empty on a duplicate/all-blocked/no-op
 * batch. GET /events/{id}/invites (listEventInvites, c351) is the paginated route
 * for "what does the whole list look like".
 */
export async function inviteToEvent(
  eventId: string,
  userIds: string[],
): Promise<EventInviteBatchOut> {
  return request<EventInviteBatchOut>(`/events/${eventId}/invites`, {
    method: "POST",
    body: { user_ids: userIds },
  });
}

// listGuests / EventGuestsOut are GONE (c275): the wrapper returned both guest
// lists unbounded. Its halves are listRsvps + listEventInvites below; headcounts
// come from getRsvpCounts, never from summing pages.

/** One page of an event's invites, earliest first. Guest-list gated. */
export async function listEventInvites(
    eventId: string,
  page: EventInvitePage = {},
): Promise<EventInviteOut[]> {
  return request<EventInviteOut[]>(`/events/${eventId}/invites`, {
    query: {
      after: page.after, after_user_id: page.afterUserId, limit: page.limit,
      unanswered_only: page.unansweredOnly,
    },
  });
}

/** Headcounts by answer plus silent invitees - the planning number (c275). */
export async function getRsvpCounts(eventId: string): Promise<EventRsvpCountsOut> {
  return request<EventRsvpCountsOut>(`/events/${eventId}/rsvp-counts`);
}

/** Events the signed-in user was invited to. Cancelled ones are included on purpose. */
export async function listMyInvites(): Promise<EventOut[]> {
  return request<EventOut[]>("/me/event-invites");
}

/**
 * One row of listMyInvitesWithRsvps() - mirrors backend EventInviteWithRsvpOut (c204).
 *
 * my_rsvp_status is the CALLER'S OWN rsvp (null if they haven't answered) - never
 * another invitee's. hosted_by is the hosting chapter's display name, resolved
 * server-side and safe to show even though the caller may not be a member of that
 * chapter - an invite already admits them to the event, which shows who hosts it.
 */
export interface EventInviteWithRsvpOut {
  event: EventOut;
  my_rsvp_status: RsvpStatus | null;
  hosted_by: string;
}

/**
 * Bulk sibling of listMyInvites() (c204): the same invited events, plus each one's
 * own-rsvp status and chapter label in a single round trip. Replaces the Home screen's
 * per-invite listGuests() + getChapter() N+1 (see feed/index.tsx's old
 * loadVisibleInvites for the shape this collapses).
 */
export interface MyInvitesPage {
  before?: string;
  beforeId?: string;
  limit?: number;
  view?: "all" | "actionable" | "history";
}

export async function listMyInvitesWithRsvps(
  page: MyInvitesPage = {},
): Promise<EventInviteWithRsvpOut[]> {
  return request<EventInviteWithRsvpOut[]>("/me/event-invites-with-rsvps", {
    query: { before: page.before, before_id: page.beforeId, limit: page.limit, view: page.view },
  });
}

/** One page of an event's RSVPs, earliest answers first. Guest-list gated. */
export async function listRsvps(
  eventId: string,
  page: GuestListPage = {},
): Promise<EventRsvpOut[]> {
  return request<EventRsvpOut[]>(`/events/${eventId}/rsvps`, {
    query: { after: page.after, after_user_id: page.afterUserId, limit: page.limit },
  });
}

export async function listRsvpPage(
  eventId: string, cursor: GuestListPage = {},
): Promise<EventPageResult<EventRsvpOut, GuestListPage>> {
  const items = await listRsvps(eventId, { ...cursor, limit: EVENT_GUEST_PAGE_SIZE });
  const last = items[items.length - 1];
  return { items, next: items.length === EVENT_GUEST_PAGE_SIZE
    ? { after: last.created_at, afterUserId: last.user_id } : null };
}

/** Only server-confirmed unanswered invitees; never subtract partial RSVP pages. */
export async function listUnansweredInvitePage(
  eventId: string, cursor: GuestListPage = {},
): Promise<EventPageResult<EventInviteOut, GuestListPage>> {
  const items = await listEventInvites(eventId, {
    ...cursor, limit: EVENT_GUEST_PAGE_SIZE, unansweredOnly: true,
  });
  const last = items[items.length - 1];
  return { items, next: items.length === EVENT_GUEST_PAGE_SIZE
    ? { after: last.created_at, afterUserId: last.invited_user_id } : null };
}

/** Home shows pending upcoming/ongoing invitations and upcoming cancellations. */
export async function listActionableInvitePage(
  cursor: MyInvitesPage = {},
): Promise<EventPageResult<EventInviteWithRsvpOut, MyInvitesPage>> {
  const items = await listMyInvitesWithRsvps({
    ...cursor, limit: HOME_INVITE_PAGE_SIZE, view: "actionable",
  });
  const last = items[items.length - 1];
  return { items, next: items.length === HOME_INVITE_PAGE_SIZE
    ? { before: last.event.starts_at, beforeId: last.event.id } : null };
}

/** Headcounts for one event - mirrors backend EventRsvpCountsOut (c275). */
export interface EventRsvpCountsOut {
  going: number;
  maybe: number;
  cant: number;
  invited_unanswered: number;
}

/** One row of listEventsWithRsvps() - mirrors backend EventWithRsvpSummaryOut (c280).
 * counts is the truth; going_preview is the first few going answers, display-sized
 * for the avatar stack and never claimed complete. */
export interface EventWithRsvpSummaryOut {
  event: EventOut;
  counts: EventRsvpCountsOut;
  going_preview: EventRsvpOut[];
  my_rsvp_status: RsvpStatus | null;
}

/** One page of the chapter's events with their RSVP summaries. `before`/`beforeId`
 * are the OLDEST event you already hold - BOTH halves or NEITHER, same rule as
 * listEventsWithRsvps' underlying (starts_at, id) cursor everywhere else in this
 * file, since `before` alone cannot tie-break two events sharing a start time. */
export interface ChapterEventsPage {
  before?: string;
  beforeId?: string;
  limit?: number;
}

/** The chapter's events with their RSVP summaries in ONE round trip (c43 shape,
 * re-cut by c280) - replaces the Events segment's listEvents + listRsvps-per-event
 * 1+N, without shipping a campus of rows per popular event.
 *
 * c359: the server has accepted before/before_id/limit on this route since c201's
 * cursor landed on list_events_with_rsvps; this client never sent them, so the
 * Events segment silently stopped at page one. */
export async function listEventsWithRsvps(
  chapterId: string,
  page: ChapterEventsPage = {},
): Promise<EventWithRsvpSummaryOut[]> {
  const paired = page.before !== undefined && page.beforeId !== undefined;
  return request<EventWithRsvpSummaryOut[]>(`/chapters/${chapterId}/events-with-rsvps`, {
    query: {
      before: paired ? page.before : undefined,
      before_id: paired ? page.beforeId : undefined,
      limit: page.limit,
    },
  });
}

/** Upserts the current user's RSVP for an event (Going / Maybe / Can't, §8.7). */
export async function setRsvp(eventId: string, status: RsvpStatus): Promise<EventRsvpOut> {
  return request<EventRsvpOut>(`/events/${eventId}/rsvps`, { method: "PUT", body: { status } });
}

File: backend/app/schemas/events.py
"""Events schemas: chapter events (Partiful-style), invites and RSVPs (c33, c198)."""

import uuid
from datetime import datetime, timezone
from typing import Literal

from pydantic import Field, field_validator, model_validator

from app.core.validation import validate_public_url
from app.schemas.base import _Schema

RsvpStatus = Literal["going", "maybe", "cant"]

# Ordered by how far the event travels, narrowest first.
#   chapter  - active members of the hosting chapter only (what events meant before c198)
#   campus   - .edu-verified students of this chapter's campus (the c88 population)
#   verified - any .edu-verified user, so a sister chapter or another school can be invited
#   public   - no account at all
EventVisibility = Literal["chapter", "campus", "verified", "public"]


# ---- events ----


class EventCreate(_Schema):
    title: str = Field(min_length=1)
    starts_at: datetime
    location: str = Field(min_length=1)
    cover_url: str = Field(min_length=1)
    description: str | None = None
    ends_at: datetime | None = None
    # DEFAULTS TO THE NARROWEST TIER. A client that omits this gets a members-only
    # event, never a world-readable one - the same reason the column's server_default is
    # 'chapter'. Widening is always an explicit act by a host.
    visibility: EventVisibility = "chapter"

    # c184 sweep: cover_url is client-supplied and written straight through to the
    # events row, the same shape of gap the card flagged in alumni.linkedin_url /
    # apply_url. http(s)-only, <= 2048 chars.
    @field_validator("cover_url")
    @classmethod
    def _validate_cover_url(cls, value: str) -> str:
        validated = validate_public_url(value)
        assert validated is not None  # field is required (min_length=1), never None
        return validated

    # A naive datetime from a client is taken as UTC, same as chapters.py and
    # core/invites.py. Without this, one naive and one aware value make the ordering
    # check below (and the tz-aware column writes) raise TypeError, which pydantic
    # does not translate into a 422 - it surfaces as a raw 500.
    @field_validator("starts_at", "ends_at")
    @classmethod
    def _assume_utc(cls, value: datetime | None) -> datetime | None:
        if value is not None and value.tzinfo is None:
            return value.replace(tzinfo=timezone.utc)
        return value

    @model_validator(mode="after")
    def _ends_after_starts(self) -> "EventCreate":
        # The CHECK constraint would also reject this, but a 422 naming the problem
        # beats an IntegrityError surfacing as a 500 to a host who fat-fingered a date.
        if self.ends_at is not None and self.ends_at <= self.starts_at:
            raise ValueError("ends_at must be after starts_at")
        return self


class EventUpdate(_Schema):
    """A host's edit. Every field optional; omitted fields are left alone.

    OMITTED VS EXPLICIT-NULL IS THE WHOLE CONTRACT HERE, and pydantic v2 already gives
    it to us for free: update_event reads body.model_dump(exclude_unset=True), and
    exclude_unset looks at model_fields_set, not at the value. A field left out of the
    JSON body entirely never enters model_fields_set and is excluded - "leave it
    alone". A field sent as explicit `null` (e.g. {"ends_at": null}) DOES enter
    model_fields_set with a value of None, survives exclude_unset, and reaches
    setattr(event, "ends_at", None) - "clear it". So {"ends_at": null} in the request
    body clears a previously-set end time (and likewise for description); leaving
    ends_at out of the body keeps whatever was stored. See test_events.py's
    test_explicit_null_clears_ends_at_and_description_omission_leaves_them (c202) for
    the proof, including that a cleared ends_at correctly skips the ends-after-starts
    check on the next edit rather than failing it.
    """

    title: str | None = Field(default=None, min_length=1)
    starts_at: datetime | None = None
    ends_at: datetime | None = None
    location: str | None = Field(default=None, min_length=1)
    cover_url: str | None = Field(default=None, min_length=1)
    description: str | None = None
    visibility: EventVisibility | None = None

    @field_validator("cover_url")
    @classmethod
    def _validate_cover_url(cls, value: str | None) -> str | None:
        return validate_public_url(value) if value is not None else None

    # Same normalization as EventCreate: update_event compares the submitted value
    # against the STORED one (always aware, the columns are timezone=True), so a naive
    # datetime here would hit the same TypeError-as-500 inside the route handler.
    @field_validator("starts_at", "ends_at")
    @classmethod
    def _assume_utc(cls, value: datetime | None) -> datetime | None:
        if value is not None and value.tzinfo is None:
            return value.replace(tzinfo=timezone.utc)
        return value


class EventOut(_Schema):
    """An event as somebody who is allowed to see it reads it."""

    id: uuid.UUID
    chapter_id: uuid.UUID
    title: str
    cover_url: str
    description: str | None = None
    starts_at: datetime
    ends_at: datetime | None = None
    location: str
    visibility: EventVisibility
    canceled_at: datetime | None = None
    host_id: uuid.UUID
    created_at: datetime


class PublicEventOut(_Schema):
    """An event as an UNAUTHENTICATED reader sees it. Deliberately narrower.

    THIS IS A SAFETY BOUNDARY, NOT A CONVENIENCE. A public event is a hole through the
    c88 .edu gate that braul opened knowingly on c198, and the only thing keeping the
    hole the size it was agreed to be is that this class carries fewer fields than
    EventOut. It has NO host_id, NO chapter_id, NO guest list and NO RSVP breakdown:
    the internet gets to know a party is happening, not who is going to it.

    Anyone "simplifying" this by returning EventOut from the public route is widening a
    decision that was made with the risk written down. The going_count is the single
    number here that is derived from other people's behaviour, and it is a count rather
    than names on purpose.
    """

    id: uuid.UUID
    title: str
    cover_url: str
    description: str | None = None
    starts_at: datetime
    ends_at: datetime | None = None
    location: str
    canceled_at: datetime | None = None
    # Display name of the hosting chapter ("Sigma Chi Epsilon Mu"), not its id - a
    # public reader has no business enumerating chapter ids.
    hosted_by: str
    going_count: int


# ---- invites ----


class EventInviteCreate(_Schema):
    """Invite people to an event. Idempotent per person by construction.

    A list rather than one id per request: inviting the roster is the ordinary case, and
    one request per member would be N round trips and N chances to half-finish.
    """

    # Bounded because the route writes one EventInvite row per id (routers/events.py's
    # pg_insert): an unbounded list lets the caller decide how much work one request
    # costs, and a rate limit caps call frequency rather than rows per call (c263/c264).
    #
    # The same 500 as MeetingAttendanceUpdate.entries, on purpose and not by coincidence:
    # both take a whole chapter roster, so "far above any real roster and far below a
    # payload that could tie up a connection" resolves to the same number. A second,
    # bespoke ceiling for the same object would only be a second thing to keep in sync.
    user_ids: list[uuid.UUID] = Field(min_length=1, max_length=500)


class EventInviteOut(_Schema):
    event_id: uuid.UUID
    invited_user_id: uuid.UUID
    invited_by: uuid.UUID
    created_at: datetime


# ---- rsvps ----


class EventRsvpUpdate(_Schema):
    status: RsvpStatus


class EventRsvpOut(_Schema):
    event_id: uuid.UUID
    user_id: uuid.UUID
    status: RsvpStatus
    created_at: datetime


class EventRsvpCountsOut(_Schema):
    """Headcounts for one event: the host's planning number (c275).

    invited_unanswered is invite rows with no rsvp row at all - an invitee who
    answered 'cant' has answered and sits in cant. Kept alongside the three status
    buckets because "invited 40, heard back from 12" is the other half of planning.
    """

    going: int
    maybe: int
    cant: int
    invited_unanswered: int


class EventWithRsvpSummaryOut(_Schema):
    """One row of GET /chapters/{id}/events-with-rsvps: c43's one-round-trip shape,
    re-cut by c280 to carry a SUMMARY instead of every RSVP row.

    The old shape returned each event's full rsvps array, unbounded - the event page
    was cursored but rows-per-event were not, and a public-tier party accrues RSVPs
    campus-wide, so one popular event shipped a campus of rows into every Events
    segment load. What the segment actually renders is a 4-face avatar stack and a
    headcount, so that is what this carries: counts (the truth, from c275's
    EventRsvpCountsOut), a bounded going_preview for the faces, and the caller's own
    answer. The full roster lives behind the paginated guest-list routes (c275).
    """

    event: EventOut
    counts: EventRsvpCountsOut
    # First GOING_PREVIEW_LIMIT 'going' rows by (created_at, user_id) ASC - enough for
    # the avatar stack with headroom. NOT claimed complete; counts carry the truth,
    # which is why this deliberate preview needs no warn_if_capped.
    going_preview: list[EventRsvpOut]
    my_rsvp_status: RsvpStatus | None


class EventInviteWithRsvpOut(_Schema):
    """One row of GET /me/event-invites-with-rsvps (c204): an invited event, plus the
    two things the mobile client used to fetch per-invite - the same shape of fix c43
    gave a chapter's own events (now EventWithRsvpSummaryOut), applied here to a
    cross-chapter invite list.

    my_rsvp_status is the CALLER'S OWN answer, never anyone else's - null means they
    have not answered. hosted_by is the hosting chapter's display name ("{org_name}
    {chapter_name}"), safe to show a non-member because an invite already admits them
    to the event, which shows who is hosting it. It is never a placeholder like the
    client's old "Another chapter" fallback: chapter_id is a NOT NULL FK on events, so
    a joined query always resolves a real chapter.
    """

    event: EventOut
    my_rsvp_status: RsvpStatus | None
    hosted_by: str


# EventGuestsOut is GONE (c275): the wrapper it shaped returned both guest lists
# unbounded. Invites and RSVPs stay SEPARATE routes rather than ever merging -
# "invited, has not answered" is a different row from "said no", and a merged list
# cannot express someone who RSVPd without an invite (constant on public tiers).
